### PR TITLE
PrintAst: improve support for restricting subsets of the AST to print

### PR DESCRIPTION
**Pull request by smowton**
* Exclude function definitions, not just their children, when excluded by configuration
* Allow excluding files
* Allow excluding comments
* Sort root files
* Tests for all of the above

### CODE REVIEW

**Comment by smowton** [comment]
@aeisenberg there is a change to that file in this PR; is it possible you saw an old diff?

**Comment by aeisenberg** [comment]
I was talking about the file added in #262, which I think would need to override `shouldPrintFile`.

**Comment by smowton** [comment]
Ah, yes indeed it should

File: ql/src/semmle/go/PrintAst.qll [hunk 1]
```diff
@@ -18,19 +22,34 @@ class PrintAstConfiguration extends string {
...
 private predicate shouldPrintFunction(FuncDef func) {
   exists(PrintAstConfiguration config | config.shouldPrintFunction(func))
 }
 
+private predicate shouldPrintFile(File file) {
+  exists(PrintAstConfiguration config | config.shouldPrintFile(file))
+}
+
+private FuncDef getEnclosingFunction(AstNode n) {
+  result = n or
```
Comment (sauyon) [inline ql/src/semmle/go/PrintAst.qll:42]: We might want to try to deal with function literals here in some way; in particular this gives two values for function literals inside of functions, since it returns both the function literal itself as well as its parent function.

My inclination would be to just do logic on the highest-level enclosing function.

File: ql/src/semmle/go/PrintAst.qll [hunk 2]
```diff
@@ -18,19 +22,34 @@ class PrintAstConfiguration extends string {
...
+  result = n.getEnclosingFunction()
+}
+
 /**
  * An AST node that should be printed.
  */
 private newtype TPrintAstNode =
   TAstNode(AstNode ast) {
-    // Do print ast nodes without an enclosing function, e.g. file headers
```
Comment (sauyon) [inline ql/src/semmle/go/PrintAst.qll:32]: Why did you remove this comment? It might do with a bit of reworking but I think it still applies.

File: ql/src/semmle/go/AST.qll [hunk 3]
```diff
@@ -47,6 +47,14 @@ class AstNode extends @node, Locatable {
   /** Gets the innermost function definition to which this AST node belongs, if any. */
   FuncDef getEnclosingFunction() { result = getParent().parentInSameFunction*() }
 
+  /**
+   * Gets the file this AST node belongs to.
+   *
+   * Compare this function, which retrieves an AST ancestor, vs. `Locatable.getFile()`,
+   * which is fed by source-location information.
+   */
+  File getEnclosingFile() { result = getParent*() }
```
[the remark this answers is not in the document]
Comment (smowton) [reply ql/src/semmle/go/AST.qll:90]: Looking at how Locatable is generated I actually don't think it can, so I'll remove this

**Comment by smowton** [comment]
Updated this PR to (a) include the requested no-comments mode, (b) update the plugin query as @aeisenberg noted and (c) apply @sauyon's suggestion, to restrict function filtering to top-level FuncDecls, as a separate commit. @sauyon please re-review; @aeisenberg please check if my change to the plugin query is ok.

Admin note: it appears the orders different files appear in the graph output are quite easy to perturb; hence there is considerable noise introduced by making sure the tests pass at each commit

**Comment by aeisenberg** [comment]
Looks great to me. When I get a chance later today, I'll run this locally.

**Comment by aeisenberg** [comment]
All right. It's working much better now. There are a few more smaller things that I'm collecting, which I will put into a different issue and I don't think they should block this PR from going through.

**Comment by sauyon** [comment]
> Admin note: it appears the orders different files appear in the graph output are quite easy to perturb; hence there is considerable noise introduced by making sure the tests pass at each commit

Hm, seems like the tests failed due to this. I would suspect this has to do with the order the files are extracted in the database, since QL is meant to be deterministic.

Would it be possible to make the AST print the files in lexicographic order?

**Comment by smowton** [comment]
Looks like they'd flipped on my local machine too probably due to applying autoformatting (it's that volatile), so re-pushed with those changes. I'll merge this if they pass, but then in parallel look at making this more reliable.

**Comment by sauyon** [comment]
I just tested this locally and it seems like it really is non-deterministic. I'll ask the QL folks.

**Comment by smowton** [comment]
@sauyon added a commit implementing File node sorting

File: ql/src/semmle/go/PrintAst.qll [hunk 4]
```diff
@@ -70,8 +102,10 @@ class PrintAstNode extends TPrintAstNode {
    * is `key`.
    */
   string getProperty(string key) {
-    key = "semmle.label" and
-    result = toString()
+    (
```
Comment (sauyon) [inline ql/src/semmle/go/PrintAst.qll:105]: I guess these parentheses were for debugging?

Comment (smowton) [reply ql/src/semmle/go/PrintAst.qll]: From a previous version, removed

**Comment by smowton** [comment]
Heads up @aeisenberg that this has landed
